### PR TITLE
Reapply #2017 (Allow shapecheck of PixelCNN++), fixing #2245

### LINKED ISSUE #2245: Recent shape checking changes result in new XLA errors
It turns out that PR #2017 breaks the following code:
```
array = np.ones(5)
print(array[:10])
```

should print `[1. 1. 1. 1. 1.]` but instead throws an error
```
Traceback (most recent call last):
  File "/Users/necula/Source/jax/jax/interpreters/xla.py", line 230, in primitive_computation
    return c.Build()
  File "/Users/necula/Source/jax/jax/lib/xla_bridge.py", line 281, in Build
    *args, **kwargs)
  File "/Users/necula/Source/jax/build/jaxlib/xla_client.py", line 734, in Build
    return Computation(self._builder.Build(), backend=backend)
RuntimeError: Invalid argument: Slice size at index 0 in gather op is out of range, must be within [0, 6), got 10.: 
```
In some sense this error is reasonable, but it is not what JAX used to do before. 
Before JAX used to sanitize indices, and produce the following JAXPR for the `gather`:
```
{ lambda b ; a.
  let c = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(0,), collapsed_slice_dims=(), start_index_map=(0,))
                  operand_shape=(5,)
                  slice_sizes=(5,) ] a b
  in c }
```

After this change, the JAXPR is:
```
{ lambda b ; a.
  let c = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(0,), collapsed_slice_dims=(), start_index_map=(0,))
                  operand_shape=(5,)
                  slice_sizes=(10,) ] a b
  in c }
```

### PR DESCRIPTION
**Pull request by juliuskunze**
fixes #2245

### CODE REVIEW

**Comment by mattjj** [comment]
Just verified that internal tests and #2245 pass.

I only ran internal tests for JAX itself, not downstream projects; I'm going to run tests for downstream projects now.

**Comment by mattjj** [comment]
(Just jotting down some notes as I go through this.)

I noticed a bug in `Poly.__rsub__`, namely that we had written `self + -other` while `self.__rsub__(other)` actually means `other - self`.

File: jax/numpy/lax_numpy.py
Comment (mattjj) [inline jax/numpy/lax_numpy.py:3324]: Btw we need to use `_any` not `any` here, since the latter is a `jax.numpy` function! Tricky.

File: jax/numpy/lax_numpy.py
Comment (mattjj) [inline jax/numpy/lax_numpy.py:3384]: This seems pretty subtle... hmm...

Comment (mattjj) [reply jax/numpy/lax_numpy.py:3384]: I'm going to leave a todo here to follow up and figure out the best way to do this.

Comment (JuliusKunze) [reply jax/numpy/lax_numpy.py:3384]: `i = lax.convert_element_type(i, index_dtype)` doesn't work for cases like `x[-1]` where `i` becomes a `Poly` representing `n-1`, breaking `convert_element_type`.

Let me know if you can come up with a better solution!

**Comment by JuliusKunze** [comment]
Good catch with `__rsub__`! It's now test covered and fixed.

**Comment by mattjj** [comment]
Thanks for the quick fix!

I read through in vim and tweaked mostly whitespace/comments in the commit I just pushed. LGTM! Let's wait for Travis to be happy, then I'll merge.

**Review by mattjj — APPROVED** [review]
LGTM! Thanks for this leap forward. It's even better than #2017.

**Comment by mattjj** [comment]
Running one more pass of downstream tests to be sure...

**Comment by mattjj** [comment]
I see a failure in [this test](https://github.com/tensorflow/probability/blob/e38aab74065a122c556460de6f17f54245ec251b/tensorflow_probability/python/distributions/mvn_tril_test.py#L507) that I must look into.

File: jax/interpreters/masking.py
Comment (shoyer) [inline jax/interpreters/masking.py:92]: We consolidated similar logic into `canonicalize_shape`/`_canonicalize_dimension` in `core.py`. Could you make use of those here?

(It also helps avoid some potential circular dependency issues.)

**Comment by mattjj** [comment]
We hit the `_slice_indices` function in that test. If I replace its body with `return idx.indices(size)` then the test passes. Seems like we've got this bug cornered... but it is concerning that this issue didn't turn up in our tests!

**Comment by mattjj** [comment]
IIUC `_slice_sizes` has logic based on [this CPython code](https://github.com/python/cpython/blob/37af21b667a9f41437b5b8e451497d7725016df5/Objects/sliceobject.c#L165).

**Comment by mattjj** [comment]
I think the riskiest thing with a PR like this is that we break existing code paths, which is unfortunately what this test failure is showing us. A way to avoid that is to be sure to leave the existing paths intact, and only add new clearly-separated paths. For example, instead of adapting the existing `_static_idx` function logic to handle both `Poly` instances and ints in one if/else chain, we can just check up-front whether there are any `Poly`s involved and if so call into a new function; if not, we fall through to the existing logic unchanged.

I'll try to make that change fo `_slice_indices`, but we should go through the PR and look for all opportunities to do that. @JuliusKunze can you help me with that?

**Comment by mattjj** [comment]
The issue was that when given these inputs:

```
idx = slice(DeviceArray(44, dtype=int64), DeviceArray(-52, dtype=int64), DeviceArray(-3, dtype=int64))

size = 2
```

We get

```
idx.indices(size) == (1, -1, -3)

_slice_indices(idx, size) ==
(DeviceArray(44, dtype=int64), DeviceArray(-50, dtype=int64), DeviceArray(-3, dtype=int64))
```

**Comment by mattjj** [comment]
(I'm a bit concerned that those are DeviceArrays at all... that seems problematic.)

**Review by mattjj — COMMENTED** [review]
Actually, before being merge-able we need to
1. make sure we're not affecting existing code paths (as a silent error almost crept through), and
2. ~make sure dimensions are never DeviceArrays.~ <-- this is a problem on master

**Comment by mattjj** [comment]
Actually, I think we want to model [this CPython code](https://github.com/python/cpython/blob/6d6508765514c7c10719478a0430f5e47c9a96ac/Objects/sliceobject.c#L372).

**Comment by shoyer** [comment]
> 2\. make sure dimensions are never DeviceArrays.

I think we ensure this already, after my change to add coercion via `canonicalize_shape` into the constructor for `ShapedArray`:
https://github.com/google/jax/blob/790d92965c881e0316876242953009ca6fc8f3eb/jax/core.py#L819

**Comment by mattjj** [comment]
Perhaps, but not in this code: my comment above shows DeviceArrays present in dimensions in a debugger.

**Comment by mattjj** [comment]
By the way, I don't have particular reason to believe that this PR introduced the sizes-as-DeviceArrays issue. It's just when we're noticing it!

**Comment by mattjj** [comment]
Ah, it's because of how `_normalize_index` is written and used here. This is an issue on master. I'll just open a separate issue for it.

**Comment by mattjj** [comment]
Okay, I pushed a fix for `_slice_sizes`, basically following that CPython function. I left the Poly case as NotImplemented, which will probably break some tests in this PR. Just wanted to commit what I had so far!

**Comment by mattjj** [comment]
@JuliusKunze could you follow up on the Poly cases (where I wrote `raise NotImplementedError  # TODO` lines)? Actually, we also need to handle when `size` is a `Poly` (or perhaps only that case?) so this placeholder logic might not be quite right.

**Comment by mattjj** [comment]
Actually, I think I got things patched up. @JuliusKunze please take a look and lmk what you think! (How the turntables: now it is you who is the reviewer!)

File: jax/interpreters/masking.py
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: Ok, nice! I just removed `to_index`, `canonicalize_shape` is now used everywhere.

**Comment by JuliusKunze** [comment]
@mattjj Thanks for implementing negative steps for slicing, I added tests to cover this. I am ok with defaulting to the builtin function for the standard case. Added more tests for polymorphic start/stop indices that required slight modifications. Rest LGTM!

File: jax/api.py
Comment (mattjj) [inline jax/api.py:32]: I think these two imports are unused. (Easy to fix later!)

File: jax/numpy/lax_numpy.py
Comment (mattjj) [inline jax/numpy/lax_numpy.py:2144]: The latter case would make `size` a `DeviceArray`, which seems like something we don't want. (We'd want to use `onp` to control types here.) @JuliusKunze but why do we need this line at all?

File: jax/random.py
Comment (mattjj) [inline jax/random.py:280]: Why do we need this change in this PR?

**Comment by mattjj** [comment]
Global presubmits look good, but there are a couple unrelated-seeming changes here (commented above) that I'd like to avoid if possible.

File: jax/api.py
Comment (JuliusKunze) [reply jax/api.py]: Removed!

File: jax/random.py
Comment (JuliusKunze) [reply jax/random.py]: We don't (relict from the random implementation taken out of this PR), now removed.

File: jax/numpy/lax_numpy.py
Comment (JuliusKunze) [reply jax/numpy/lax_numpy.py]: @mattjj According to the [XLA spec](https://www.tensorflow.org/xla/operation_semantics#iota) `lax.iota` is not supposed to accept `float` size, so I added conversion here and made `iota` more strict. 

In order to minimize the diff, I reverted the change for now and put the conversion to `int` to [where it was before](https://github.com/google/jax/pull/2800/files#diff-08adae3ce53df47932e06971edb68642R1237) (otherwise tests fail since arange accepts float). (`arange`/`iota` currently doesn't work with `shapecheck` anyway. That will be part of the next PR, so no need to check for `Poly` here for now.)

**Comment by mattjj** [comment]
Internal tests green!

**Pull request merged**
